cns/service.go

```diff
@@ -48,62 +56,85 @@ func NewService(name, version, channelMode string, store store.KeyValueStore) (*
 	}, nil
 }
 
-// GetAPIServerURL returns the API server URL.
-func (service *Service) getAPIServerURL() string {
-	urls, _ := service.GetOption(acn.OptCnsURL).(string)
-	if urls == "" {
-		urls = defaultAPIServerURL
-	}
-
-	return urls
-}
-
-// Initialize initializes the service and starts the listener.
-func (service *Service) Initialize(config *common.ServiceConfig) error {
-	log.Debugf("[Azure CNS] Going to initialize a service with config: %+v", config)
-
-	// Initialize the base service.
-	if err := service.Service.Initialize(config); err != nil {
-		return errors.Wrap(err, "failed to initialize")
+// AddListeners is to add two listeners(nodeListener and localListener) for connections on the given address
+func (service *Service) AddListeners(config *common.ServiceConfig) error {
+	// Fetch and parse the API server URL.
+	var nodeURL *url.URL
+	cnsURL, _ := service.GetOption(acn.OptCnsURL).(string)
+	if cnsURL == "" {
+		// get VM primary interface's private IP
+		nodeURL, _ = url.Parse(fmt.Sprintf("tcp://%s:%s", config.PrimaryInterfaceIP, defaultAPIServerPort))
+	} else {
+		// use the URL that customer provides
+		nodeURL, _ = url.Parse(cnsURL)
 	}
 
-	// Initialize the listener.
-	if config.Listener == nil {
-		// Fetch and parse the API server URL.
-		u, err := url.Parse(service.getAPIServerURL())
-		if err != nil {
-			return err
-		}
+	cnsListener := cnsListener{}
 
-		listener, err := acn.NewListener(u)
+	if config.ChannelMode != CRD {
+		// construct url
+		nodeListener, err := acn.NewListener(nodeURL)
 		if err != nil {
-			return err
+			return errors.Wrap(err, "Failed to construct url for node listener")
 		}
 
-		if config.TlsSettings.TLSPort != "" {
+		// only use TLS connection for DNC/CNS listener:
+		if config.TLSSettings.TLSPort != "" {
 			// listener.URL.Host will always be hostname:port, passed in to CNS via CNS command
 			// else it will default to localhost
 			// extract hostname and override tls port.
-			hostParts := strings.Split(listener.URL.Host, ":")
-			tlsAddress := net.JoinHostPort(hostParts[0], config.TlsSettings.TLSPort)
+			hostParts := strings.Split(nodeListener.URL.Host, ":")
+			tlsAddress := net.JoinHostPort(hostParts[0], config.TLSSettings.TLSPort)
 
 			// Start the listener and HTTP and HTTPS server.
-			tlsConfig, err := getTLSConfig(config.TlsSettings, config.ErrChan)
+			tlsConfig, err := getTLSConfig(config.TLSSettings, config.ErrChan) //nolint
 			if err != nil {
 				log.Printf("Failed to compose Tls Configuration with error: %+v", err)
 				return errors.Wrap(err, "could not get tls config")
 			}
 
-			if err := listener.StartTLS(config.ErrChan, tlsConfig, tlsAddress); err != nil {
-				return err
+			if err := nodeListener.StartTLS(config.ErrChan, tlsConfig, tlsAddress); err != nil {
+				return errors.Wrap(err, "could not start tls")
 			}
 		}
+		config.Listeners = append(config.Listeners, nodeListener)
```

Comment:
why do we need another array of Listeners `config.Listeners`? can we just use `service.Listeners`

Reply:
This was the part of cnsconfig struct field in there but yes we do not need it